base/strings/util.jl

```diff
@@ -633,30 +633,36 @@ julia> hex2bytes(a)
 function hex2bytes end
 
 hex2bytes(s::AbstractString) = hex2bytes(String(s))
-hex2bytes(s::Union{String,AbstractVector{UInt8}}) = hex2bytes!(Vector{UInt8}(undef, length(s) >> 1), s)
-
-_firstbyteidx(s::String) = 1
-_firstbyteidx(s::AbstractVector{UInt8}) = first(eachindex(s))
-_lastbyteidx(s::String) = sizeof(s)
-_lastbyteidx(s::AbstractVector{UInt8}) = lastindex(s)
+hex2bytes(s::String) = hex2bytes(transcode(UInt8, s))
+hex2bytes(s) = hex2bytes!(Vector{UInt8}(undef, length(s) >> 1), s)
 
 """
-    hex2bytes!(d::AbstractVector{UInt8}, s::Union{String,AbstractVector{UInt8}})
+    hex2bytes!(dest::AbstractVector{UInt8}, itr)
 
-Convert an array `s` of bytes representing a hexadecimal string to its binary
+Convert an iterable `itr` of bytes representing a hexadecimal string to its binary
 representation, similar to [`hex2bytes`](@ref) except that the output is written in-place
-in `d`.   The length of `s` must be exactly twice the length of `d`.
-"""
-function hex2bytes!(d::AbstractVector{UInt8}, s::Union{String,AbstractVector{UInt8}})
-    if 2length(d) != sizeof(s)
-        isodd(sizeof(s)) && throw(ArgumentError("input hex array must have even length"))
-        throw(ArgumentError("output array must be half length of input array"))
+to `dest`. The length of `dest` must be at least half the length of `itr`.
+"""
+function hex2bytes!(dest::AbstractArray{UInt8}, itr)
+    isodd(length(itr)) && throw(ArgumentError("length of iterable must be even"))
+    @boundscheck 2*length(dest) < length(itr) && throw(ArgumentError("length of output array must be at least half of the length of input array"))
```

Comment:
I think this change will cause the bug with uninitialized elements.

Reply:
You mean if `dest` is bigger than half the length of `itr`? In that case, the user had provided it uninitialized. The forward from `hex2bytes(itr)` above creates `dest` of the correct size.

Reply:
`hex2bytes!` is a public API. Risky specification changes should be avoided.

Reply:
I'm not sure if it's a risky change - so far, the length had to be exactly half of the source string/vector and larger vectors were an error. I've also changed it to `AbstractArray` to allow for other dimensionalities (and their views).

I can change it back to `AbstractVector` and `2*length(dest) == length(itr)` though if it's a big problem.

Reply:
I also don't think this is a big problem. However, bugs arising from trivial problems are more troublesome.
I respect the opinion of the author (and of course, the community).

Reply:
I've changed it back to the inequality - I've thought about this and I've come to the conclusion that it's better for the user to be explicit about only passing part of an uninitialized array to this function.